src/csharp/ChatClient.cs

```diff
@@ -0,0 +1,206 @@
+            generatorParams.SetInputSequences(tokens);
+
+            using Generator generator = new(_model, generatorParams);
+            using Sequences outputSequences = _model.Generate(generatorParams);
+
+            return new ChatCompletion(new ChatMessage(ChatRole.Assistant, _tokenizer.Decode(outputSequences[0])))
+            {
+                CompletionId = Guid.NewGuid().ToString(),
+                CreatedAt = DateTimeOffset.UtcNow,
+                ModelId = Metadata.ModelId,
+            };
+        }, cancellationToken);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
+        IList<ChatMessage> chatMessages, ChatOptions options = null, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (chatMessages is null)
+        {
+            throw new ArgumentNullException(nameof(chatMessages));
+        }
+
+        using Sequences tokens = _tokenizer.Encode(CreatePrompt(chatMessages));
+        using GeneratorParams generatorParams = new(_model);
+        UpdateGeneratorParamsFromOptions(tokens[0].Length, generatorParams, options);
+        generatorParams.SetInputSequences(tokens);
+
+        using Generator generator = new(_model, generatorParams);
+        using var tokenizerStream = _tokenizer.CreateStream();
+
+        var completionId = Guid.NewGuid().ToString();
+        while (!generator.IsDone())
+        {
+            string next = await Task.Run(() =>
+            {
+                generator.ComputeLogits();
+                generator.GenerateNextToken();
+
+                ReadOnlySpan<int> outputSequence = generator.GetSequence(0);
+                return tokenizerStream.Decode(outputSequence[outputSequence.Length - 1]);
+            }, cancellationToken);
+
+            yield return new StreamingChatCompletionUpdate
+            {
+                CompletionId = completionId,
+                CreatedAt = DateTimeOffset.UtcNow,
+                Role = ChatRole.Assistant,
+                Text = next,
+            };
+        }
+    }
+
+    /// <inheritdoc/>
+    public TService GetService<TService>(object key = null) where TService : class =>
+        typeof(TService) == typeof(Model) ? (TService)(object)_model :
+        typeof(TService) == typeof(Tokenizer) ? (TService)(object)_tokenizer :
+        this as TService;
+
+    /// <summary>Creates a prompt string from the supplied chat history.</summary>
+    private string CreatePrompt(IEnumerable<ChatMessage> messages)
+    {
+        StringBuilder prompt = new();
+
+        foreach (var message in messages)
+        {
+            foreach (var content in message.Contents)
+            {
+                switch (content)
+                {
+                    case TextContent tc when !string.IsNullOrWhiteSpace(tc.Text):
+                        prompt.Append("<|").Append(message.Role.Value).Append("|>\n").Append(tc.Text);
```

Comment:
Should make this more general to support non phi models. Mistral and llama for example have a different prompt template

Reply:
We can parameterize the client with the template to use. Is that information encoded anywhere in a programmatic way? Or developers with ONNX Runtime GenAI today are supposed to look up the format somewhere?

Reply:
Yeah, unfortunately, there isn't a programmatic way. But I believe you are working on handling this with @azchohfi offline already